explainers/beforematch.md

```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
 5. **content-visibility: hidden-matchable -> visible**: find-in-page scrolls to
 to the revealed text.
 
+### Alternatives
+Given the purpose of displaying `content-visibility: hidden-matchable` text
+when it is searched for, there are a number of alternatives we have considered.
+
+#### Automatic Unlocking
+`content-visibility: hidden-matchable` text would be automatically changed to
+`content-visibility: visible` when the hidden text is searched for.
+##### Pros
+* Provides the desired behavior without the need for the web developer to do
+  anything besides use the `hidden-matchable` value.
+* Less complex than `beforematch` with regards to scrolling behavior.
+##### Cons
+* Doesn't allow the developer to change other properties in conjunction with
+  displaying hidden content. For example, in "Example 1: Expanding
+  `hidden-matchable`," automatic unlocking wouldn't be able to change the
+  style of the `<h1>` title elements.
+
+#### Automatic Unlocking with Activation Event
+This is like "Automatic Unlocking," but with an added "activation" emitted when
+content is displayed to allow developers to change other state and styles if
+needed.
+##### Pros
+* Same pros as "Automatic Unlocking."
+* Allows other state/styles to be changed when content is displayed.
+##### Cons
+* TODO: ask vmpstr and chrishtr for more context on this
```

Comment:
What was the reason not to go with activation events again?

Reply:
So it depends how you look at it
- On the one hand, activation event goes hand-in-hand with automatic unlocking/expanding, which we think is not a good use-case since the majority of the time the developer would want to do something other than just show the element. That, and it makes state tracking hard (ie the developer doesn't have the tools to collapse the section again; we would have to add something like a JS api). In this case activation event informs that this has happened. However, since we're not doing the automatic unlocking then activation event is not needed.

- On the other hand, activation event could be fired prior to us finding a match that would have otherwise unlocked the element. That's exactly what beforematch is :) In other words, we renamed activation event to beforematch when we decided not to automatically unlock due to reasoning above.

Does that make sense?

Reply:
Just to add to this as I'm reading the doc, the important con is that automatically changing CSS isn't possible, so there is a lot of complexity involved with automatic unlocking. And the whole point of this activation event here is that if we're firing an activation event anyway, then why bother with automatic unlocking? Just let the script take care of it and reduce the complexity and internal states in the browser.

To put it differently, automatic unlocking with an event is not any better than just the event, because the handler in the event could still remove the element, relock the element, move it, etc, etc. All of those need to be dealt with whether or not we automatically unlock in the presence of an event.

Reply:
Why isn't automatically changing CSS possible? If we are able to fire an event that has the power to change CSS, then can't we just change the style from `hidden-matchable` to `visible` in C++?

Reply:
Which style do we change? We can't change the rule style because the rule applies to more than one element (in the general case; ie * { content-visibility: hidden-matchable } applies to _all_ elements, so when we activate, we can't just change that to visible since that un-hides everything). 

The only possibility is that we set the inline style of the element to have content-visibility: visible. But again, if for example the descendant elements have content-visibility: inherit or something then we're changing more than just the one element. That, and changing the inline style automatically open the door to a lot of web breaking, since if script generates and sets a new inline style, all of a sudden it overrides the values we set. That's just not going to be a good change.

The idea of the event is to defer this decision to script, since there is no one correct decision in terms of what to do, so we have to delegate the decision to the app.

Reply:
Ah I see, having the browser change styles doesn't really make sense. Thanks for the explanation.
I added another alternative to show why this is a bad idea.